// @astrojs/language-server/dist/core/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstroFile = exports.getLanguageModule = void 0;
const language_core_1 = require("@volar/language-core");
const path = __importStar(require("node:path"));
const astro2tsx_1 = require("./astro2tsx");
const parseAstro_1 = require("./parseAstro");
const parseCSS_1 = require("./parseCSS");
const parseHTML_1 = require("./parseHTML");
const parseJS_js_1 = require("./parseJS.js");
function getLanguageModule(astroInstall, ts) {
    return {
        createVirtualFile(fileName, snapshot) {
            if (fileName.endsWith('.astro')) {
                return new AstroFile(fileName, snapshot, ts);
            }
        },
        updateVirtualFile(astroFile, snapshot) {
            astroFile.update(snapshot);
        },
        resolveHost(host) {
            return {
                ...host,
                resolveModuleName(moduleName, impliedNodeFormat) {
                    if (impliedNodeFormat === ts.ModuleKind.ESNext &&
                        (moduleName.endsWith('.astro') ||
                            moduleName.endsWith('.vue') ||
                            moduleName.endsWith('.svelte'))) {
                        return `${moduleName}.js`;
                    }
                    return host.resolveModuleName?.(moduleName, impliedNodeFormat) ?? moduleName;
                },
                getScriptFileNames() {
                    const fileNames = host.getScriptFileNames();
                    return [
                        ...fileNames,
                        ...(astroInstall
                            ? ['./env.d.ts', './astro-jsx.d.ts'].map((filePath) => ts.sys.resolvePath(path.resolve(astroInstall.path, filePath)))
                            : []),
                    ];
                },
                getCompilationSettings() {
                    const baseCompilationSettings = host.getCompilationSettings();
                    return {
                        ...baseCompilationSettings,
                        module: ts.ModuleKind.ESNext ?? 99,
                        target: ts.ScriptTarget.ESNext ?? 99,
                        jsx: ts.JsxEmit.Preserve ?? 1,
                        jsxImportSource: undefined,
                        jsxFactory: 'astroHTML',
                        resolveJsonModule: true,
                        allowJs: true,
                        isolatedModules: true,
                        moduleResolution: baseCompilationSettings.moduleResolution === ts.ModuleResolutionKind.Classic ||
                            !baseCompilationSettings.moduleResolution
                            ? ts.ModuleResolutionKind.Node10
                            : baseCompilationSettings.moduleResolution,
                    };
                },
            };
        },
    };
}
exports.getLanguageModule = getLanguageModule;
class AstroFile {
    constructor(sourceFileName, snapshot, ts) {
        this.sourceFileName = sourceFileName;
        this.snapshot = snapshot;
        this.ts = ts;
        this.kind = language_core_1.FileKind.TextFile;
        this.capabilities = language_core_1.FileCapabilities.full;
        this.codegenStacks = [];
        this.fileName = sourceFileName;
        this.onSnapshotUpdated();
    }
    get hasCompilationErrors() {
        return this.compilerDiagnostics.filter((diag) => diag.severity === 1).length > 0;
    }
    update(newSnapshot) {
        this.snapshot = newSnapshot;
        this.onSnapshotUpdated();
    }
    onSnapshotUpdated() {
        this.mappings = [
            {
                sourceRange: [0, this.snapshot.getLength()],
                generatedRange: [0, this.snapshot.getLength()],
                data: language_core_1.FileRangeCapabilities.full,
            },
        ];
        this.astroMeta = (0, parseAstro_1.getAstroMetadata)(this.snapshot.getText(0, this.snapshot.getLength()));
        const { htmlDocument, virtualFile: htmlVirtualFile } = (0, parseHTML_1.parseHTML)(this.fileName, this.snapshot, this.astroMeta.frontmatter.status === 'closed'
            ? this.astroMeta.frontmatter.position.end.offset
            : 0);
        this.htmlDocument = htmlDocument;
        const scriptTags = (0, parseJS_js_1.extractScriptTags)(this.fileName, this.snapshot, htmlDocument, this.astroMeta.ast);
        this.scriptFiles = scriptTags.map((scriptTag) => scriptTag.fileName);
        htmlVirtualFile.embeddedFiles.push(...(0, parseCSS_1.extractStylesheets)(this.fileName, this.snapshot, htmlDocument, this.astroMeta.ast), ...scriptTags);
        this.embeddedFiles = [];
        this.embeddedFiles.push(htmlVirtualFile);
        const tsx = (0, astro2tsx_1.astro2tsx)(this.snapshot.getText(0, this.snapshot.getLength()), this.fileName, this.ts, htmlDocument);
        this.compilerDiagnostics = tsx.diagnostics;
        this.embeddedFiles.push(tsx.virtualFile);
    }
}
exports.AstroFile = AstroFile;
